### ISSUE TITLE
Items fly off screen in kanban mode

### ISSUE THREAD

**Issue by Zireael** [haltu/muuri#325]
Hi,
Using the kanban demo to show the issue:
https://codepen.io/niklasramo/pen/BrYaOp

A video can be seen here:
https://streamable.com/chev0

Here's better view using dropzone placeholders:
https://streamable.com/dolt6

The issue is that if the item drop zone isn't aligned perfectly, Muuri assigns the drop zone at the end of stack and the item flies off screen to the end of stack.

I realize the primary purpose of Muuri is 2D grid, but I'd like to use it for 1D kanban lists. Can't find a way to prevent the flying of item to the end of stack. Would it be possible to add a 1D setting mode where an item 'snaps' above or under (or left/right in horizontal mode) to the nearest item instead of being sent to end of stack?

**Comment by niklasramo** [comment]
@Zireael this issue has been fixed in the dev branch and will be released with the next version 0.9.0: https://github.com/haltu/muuri/issues/298

Although, I don't know yet _when_ that release will be.

**Comment by Zireael** [comment]
Thanks for the answer. I might have to look into the dragSortPredicate as you say.
After playing with the borked demo, I think that the item flies to the end of the stack when the grid decides it no longer belongs to the old stack, but can't figure out where does it sit in the new stack.

Few ideas here:
1) Make the behavior more conservative: if the position in the new list can't be discerned, return the item to the old list
2) Create an invisible force field around the item and compute it's target zone a bit ahead of it actually getting there (in the fly demo it seems the items fly when I release them on the side of a list). Also, it felt like if I drag items very fast, the code can't keep up with item position. I have `dragSortInterval: 0`, but maybe there is some kind of internal debounce function you're using to not fire drag event coordinate computing every animation frame?
3) Apply a logic like this: on `dragend`/`mouseup` calculate which other Muuri items does the item overlap with (% area for each item). Then, for the item that overlaps the most, determine from which side does the dragged item arrive at the target element (i.e. which edges of the dragged element and of the target element are covered the most by each other). Then drop the dragged item in place of the target item, pushich the target item to the side from where the dragged item arrives.

I'm not yet very familiar with your library's code/workings, but I hope any of this helps. Just  bit of blue sky thinking.

Thanks!

**Comment by niklasramo** [comment]
@Zireael Thanks for the ideas, been thinking some them myself too.

At the moment when you drag an item the dragged item must overlap with another item in order for us to determine a new position for the item. This approach is a bit naive and limited, and the main reason why we can't drop items on holes in the grid. However, this approach also works decently for _any_ kind of layout algorithm. If we could come up with a generic logic/algorithm on how to drop items on holes in the grid I believe this issue you're facing would also disappear, and that's the ultimate goal here :) I just haven't figured out a good solution to it yet...

> 1. Make the behavior more conservative: if the position in the new list can't be discerned, return the item to the old list

Yep, this could be one approach and I've made an exception in the normal logic when dragging item between grids. It just made sense when I tested this and compared the two approaches.

> 2. Create an invisible force field around the item and compute it's target zone a bit ahead of it actually getting there (in the fly demo it seems the items fly when I release them on the side of a list). Also, it felt like if I drag items very fast, the code can't keep up with item position. I have dragSortInterval: 0, but maybe there is some kind of internal debounce function you're using to not fire drag event coordinate computing every animation frame?

I believe what you're experiencing here is the what I've fixed in #298 . The drag position is internally stored _immediately_ when item is moved, but the DOM changes are applied asynchronously in the next frame. Although, the DOM changes are done async they still use the latest values retrieved from the touch/mouse/pointer events.

> 3. Apply a logic like this: on dragend/mouseup calculate which other Muuri items does the item overlap with (% area for each item). Then, for the item that overlaps the most, determine from which side does the dragged item arrive at the target element (i.e. which edges of the dragged element and of the target element are covered the most by each other). Then drop the dragged item in place of the target item, pushich the target item to the side from where the dragged item arrives.

The problem with this approach is that since we might have any kind of custom layout algorithm in use we don't know where to place the item based on this data. This is why it usually makes sense to build accompanying dragSortPredicate for custom layout algorithm. Most likely we'll end up something like this for the default layout algorithm _only_ and fallback to the current naive version if the default layout algorithm is not used. But we would probably do it in the normal checking flow (not only on dragend/mouseup).

> Did you ever have a play with recycle scrolling or infinite lists? Would that work for Muuri?

Nope, haven't played around with virtualization yet, but that's on the roadmap. It's going to be one tough nut to crack for sure. It's good to bear in mind that Muuri has APIs for sorting, filtering and custom layouts which all have to play nice with the virtualization solution. Good ideas and code is welcome for that part ;)
